### ISSUE TITLE
[meta-issue] Progress Tracker

### ISSUE THREAD

**Issue by pradyunsg** [pradyunsg/Py2C#11]
This is a meta-issue for tracking the progress of the project till it's first public release.
I'll keep posting updates here.
- [ ] Make a clear plan of Action for the project. (Replace this list)
- [ ] Create a (high-level) list for things to be done. Add items to it when you get to an item.

**Comment by AraHaan** [comment]
oh I did add you as a Contrib to my repo to but I am not sure if py2c would work because it needs to be able to compile to all versions of any C compiler. hopefully this can be a piece of cake for ppl that know both python and C/C++.

**Comment by pradyunsg** [comment]
@AraHaan 

1. It's not really nice to add someone as a contributor to your project without even asking them at first. Not especially when they don't know who you are or that the project even existed.

2. Moreover, saying something along the lines of 'your project is not going to work. My version is better, Work with me' to someone is not the best way of convincing them to help. Especially when ill-informed.

  From the first line in the README.md file [emphasis added]:
  > A trans-compiler for compiling Python code into human-readable **C++** code

Also, your comment is a bit of noise for this issue thus I'm going to delete this as well as your message on 16th. We can talk over my mail if you wish to. Find it in the setup.py file in this repository.

**Comment by pradyunsg** [comment]
To break the inactivity of this project and try to justify to the 100 people who probably know this project exists, as to why this project is inactive, I'm writing this. I'm pretty sure this is gonna be long.

tldr; This project is now dormant, I'm wiser, I am now looking into a different approach to the problem, mainly due to lack of time.

Q. **Why has this project gone into a dormant state?**
A. I am stuck in the vortex of Examinations for the past 6 months. I haven't really been able to put in much time to code in anything as such. Moreover, I may not get the required amount of time to invest in this project in the near future. By that I mean at-least another year or more based on how life goes. Because academic systems really care about how well you write on that sheet of paper. (that rant is for my non-existent blog, some other day).

Q. **So, no one will fix this?**
A. Through the loads of reseach/experiments I have done, in the past year or so, one thing I have come to realize is that this problem isn't exaclty an "interesting puzzle to solve" problem that I (naively) thought it was. It turns out this is a "genuinely difficult, maddeningly complicated, really hard to get working right" problem. While that most certainly doesn't mean that I'm no longer interested in solving this problem, I have a feeling I'm going to take a different approach to solving this problem.

Q. **Where is this project headed?**
A. Because of this lack of time, If it wasn't obvious by now, this project is really in a dormant state. I'm going to put a pointer in the README saying that.

---

Anyway, since I don't have access to my development machine, I'm writing down a bunch of things down.

- The current approach I'm using, of converting the Python code into an internal representation followed by a pipeline of transformations to get a representation that can then be converted into C code can be made functional, it's just a lot of work. A lot of interesting work. But a lot of work. That originates from the problem being really complicated.
- The Type-Systems of C++ and Python are very different. It is difficult even for humans at times to convert from one to another, for various reasons. Automating it will never be perfect. NEVER. But we can go close. Again, lot of work.
- It would be a bad, really bad idea to not utilize Python's really well implemented data-structures. They are the result of a ton of work by really smart people.
- There's more but now, I need to get back to my study table.

**Comment by pradyunsg** [comment]
+1

**Comment by xoviat** [comment]
@pradyunsg In case you are interested, Nuitka already does most of this.

**Comment by pradyunsg** [comment]
Hi!

It was one of the projects I was tracking and indeed it does most (if not all) of this. I kinda wanted to see to what degree I can implement these things, on my own - mostly for learning.

I am still interested in this but right now, this project is a 'nice if I complete it'. I have other things in the pipeline that need to be done first.

**Comment by Sires0** [comment]
Wanna make an IRC or Matrix room to talk? I'm really interested into a project for compiling python to raw instructions instead of using the CPython interpreter library. I have quite some free time to help.

Also what do you think of using LLVM(we could try using pyllvm) instead of C?
